### PR TITLE
Added implementation to support EventHubBufferedProducerClient

### PR DESCRIPTION
**Pull request by WhitWaldo**
Implemented my own ask from #3832 to support the EventHubBufferedProducerClient in addition to the EventHubProducerClient as part of the Aspire DI registration.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3833)

### CODE REVIEW

**Comment by eerhardt** [comment]
@WhitWaldo - it looks like the tests are failing. Can you fix them?

**Comment by eerhardt** [comment]
@sebastienros - can you take a look at this?

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return ((IAzureClientFactoryBuilderWithCredential) azureFactoryBuilder)
+            .RegisterClientFactory<EventHubBufferedProducerClient, EventHubBufferedProducerClientOptions>(
+                (options, cred) =>
+                {
+                    EnsureConnectionStringOrNamespaceProvided(settings, connectionName, configurationSectionName);
+
+                    return !string.IsNullOrEmpty(settings.ConnectionString)
+                        ? new EventHubBufferedProducerClient(settings.ConnectionString, options)
+                        : new EventHubBufferedProducerClient(settings.Namespace, settings.EventHubName, cred, options);
```
Comment (sebastienros) [inline src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:38]: `FullyQualifiedNamespace` probably. There was a recent renaming, maybe you are using an older reference.

**Comment by skolima** [comment]
👋 @WhitWaldo - I'm looking at the same piece of functionality missing. Would you intend to bring this up to date, or are you ok with someone else updating it?

**Comment by WhitWaldo** [comment]
@skolima I can bring it up to date over the weekend. It just fell off my radar!

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs (hunk 1 above)
Comment (WhitWaldo) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs]: You're right - fixed in the recent commits.

**Comment by WhitWaldo** [comment]
This has been updated and should be ready to go now!

**Comment by eerhardt** [comment]
@sebastienros - can you take a look?

File: src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs [hunk 2]
```diff
@@ -156,6 +156,54 @@ public static void AddKeyedAzureEventHubProducerClient(
...
+    /// <param name="configureSettings">An optional method that can be used for customizing the <see cref="AzureMessagingEventHubsBufferedProducerSettings"/>. It's invoked after the settings are read from the configuration.</param>
+    /// <remarks>Reads the configuration from "Aspire:Azure:Messaging:EventHubs:{TClient}" section, where {TClient} is the type of Event Hubs client being configured, i.e. EventProcessorClient.</remarks>
+    /// <param name="configureClientBuilder">An optional method that can be used for customizing the <see cref="IAzureClientBuilder{TClient, TOptions}"/>.</param>
+    /// <remarks>Reads the configuration from "Aspire:Azure:Messaging:EventHubs:{TClient}" section, where {TClient} is the type of Event Hubs client being configured, i.e. EventProcessorClient.</remarks>
+    /// <exception cref="InvalidOperationException">Thrown when neither <see cref="AzureMessagingEventHubsSettings.ConnectionString"/> nor <see cref="AzureMessagingEventHubsSettings.FullyQualifiedNamespace"/> is provided.</exception>
```
Comment (radical) [inline src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:190]: Add `<exception` for `ArgumentException` thrown for `name`.

File: src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs [hunk 3]
```diff
@@ -156,6 +156,54 @@ public static void AddKeyedAzureEventHubProducerClient(
...
+    public static void AddKeyedAzureEventHubBufferedProducerClient(
```
Comment (radical) [inline src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:191]: Just curious - when do we return the builder from such a method, and when do we use void?
cc @sebastienros @eerhardt

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+{
+    protected override void BindSettingsToConfiguration(AzureMessagingEventHubsBufferedProducerSettings settings,
+        IConfiguration configuration)
+    {
+        configuration.Bind(settings);
+    }
+
+    protected override void BindClientOptionsToConfiguration(IAzureClientBuilder<EventHubBufferedProducerClient, EventHubBufferedProducerClientOptions> clientBuilder, IConfiguration configuration)
+    {
+#pragma warning disable IDE0200 // Remove unnecessary lambda expression - needed so the ConfigBinder Source Generator works
```
Comment (radical) [inline src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:22]: Just curious - what is this about? Is it a bug in the source generator that we are working around?

File: src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs (hunk 2 above)
Comment (WhitWaldo) [reply src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:190]: Added

File: src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs (hunk 3 above)
Comment (WhitWaldo) [reply src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:191]: I'll leave that philosophy to others, but here I'm just mirroring the return type taken for the existing extensions (`void)`.

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs (hunk 4 above)
Comment (WhitWaldo) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:22]: I simply copied the approach taken in the existing [consumer client component](https://github.com/WhitWaldo/dotnet-aspire/blob/2e968b6ae9771f294aca866f94179b176599ba00/src/Components/Aspire.Azure.Messaging.EventHubs/EventHubConsumerClientComponent.cs#L17). I don't know what it is it's referring to, but if it's necessary there, I figured it should be included in this class as well.

**Comment by WhitWaldo** [comment]
All Aspire.Azure.Messaging.EventHubs.Tests are passing on my system and I believe I've spoken to each of the reviews above.

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs (hunk 4 above)
Comment (sebastienros) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:22]: `configuration.Bind` is replaced at compiled time by a source generated version. But if you use a method group then the source generator doesn't kick in. To see the code expand the project like this:

image

Look for a method named `Bind_EventHubBufferedProducerClientOptions` and check the attribute on it that will point to the method to switch for.

**Comment by sebastienros** [comment]
I think the only thing missing is updating the README to list the new client/settings. Just do a quick search for ProducerClient for the missing mentions.

**Comment by sebastienros** [comment]
@WhitWaldo I did the required changes in the README.

**Review by sebastienros — APPROVED** [review]

**Comment by radical** [comment]
Multiple tests failing like this ([log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-3833-merge-0f4c0266e13343709d/Aspire.Azure.Messaging.EventHubs.Tests/1/console.63d9478e.log?helixlogtype=result)):
```
Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.CanAddMultipleKeyedServices(clientIndex: 2) [FAIL]
  System.InvalidOperationException : No service for type 'Azure.Messaging.EventHubs.EventProcessorClient' has been registered.
  Stack Trace:
       at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetRequiredKeyedService(Type serviceType, Object serviceKey, ServiceProviderEngineScope serviceProviderEngineScope)
       at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetRequiredKeyedService(Type serviceType, Object serviceKey)
       at Microsoft.Extensions.DependencyInjection.ServiceProviderKeyedServiceExtensions.GetRequiredKeyedService(IServiceProvider provider, Type serviceType, Object serviceKey)
    /_/tests/Aspire.Azure.Messaging.EventHubs.Tests/AspireEventHubsExtensionsTests.cs(259,0): at Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.RetrieveClient(Object key, Int32 clientIndex, IHost host)
    /_/tests/Aspire.Azure.Messaging.EventHubs.Tests/AspireEventHubsExtensionsTests.cs(361,0): at Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.CanAddMultipleKeyedServices(Int32 clientIndex)
       at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
       at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.CanAddMultipleKeyedServices(clientIndex: 3) [FAIL]
  System.InvalidOperationException : No service for type 'Azure.Messaging.EventHubs.Primitives.PartitionReceiver' has been registered.
  Stack Trace:
       at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetRequiredKeyedService(Type serviceType, Object serviceKey, ServiceProviderEngineScope serviceProviderEngineScope)
       at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetRequiredKeyedService(Type serviceType, Object serviceKey)
       at Microsoft.Extensions.DependencyInjection.ServiceProviderKeyedServiceExtensions.GetRequiredKeyedService(IServiceProvider provider, Type serviceType, Object serviceKey)
    /_/tests/Aspire.Azure.Messaging.EventHubs.Tests/AspireEventHubsExtensionsTests.cs(259,0): at Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.RetrieveClient(Object key, Int32 clientIndex, IHost host)
    /_/tests/Aspire.Azure.Messaging.EventHubs.Tests/AspireEventHubsExtensionsTests.cs(361,0): at Aspire.Azure.Messaging.EventHubs.Tests.AspireEventHubsExtensionsTests.CanAddMultipleKeyedServices(Int32 clientIndex)
       at InvokeStub_AspireEventHubsExtensionsTests.CanAddMultipleKeyedServices(Object, Span`1)
       at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

**Comment by WhitWaldo** [comment]
> Multiple tests failing like this ([log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-3833-merge-0f4c0266e13343709d/Aspire.Azure.Messaging.EventHubs.Tests/1/console.63d9478e.log?helixlogtype=result)):

Out of curiosity, is there any way to access that log without being a part of Microsoft? The tests passed locally but I saw the bot report a failure and I couldn't figure out how to access the log myself.

**Comment by radical** [comment]
> > Multiple tests failing like this ([log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-3833-merge-0f4c0266e13343709d/Aspire.Azure.Messaging.EventHubs.Tests/1/console.63d9478e.log?helixlogtype=result)):
> 
> Out of curiosity, is there any way to access that log without being a part of Microsoft? The tests passed locally but I saw the bot report a failure and I couldn't figure out how to access the log myself.

Are you not able to access that link? I pasted the failures here anyway - https://gist.github.com/radical/1f55987361a1a828b3ed888b081e2dbe .

**Comment by WhitWaldo** [comment]
When I ran the tests locally, I believe I ran them against the wrong branch. Re-resolved the comments above on the correct branch this time and fixed the unit tests + added inline tests for the new class to the mix. _Now_, they're all passing locally.

**Comment by WhitWaldo** [comment]
> > > Multiple tests failing like this ([log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-3833-merge-0f4c0266e13343709d/Aspire.Azure.Messaging.EventHubs.Tests/1/console.63d9478e.log?helixlogtype=result)):
> > 
> > 
> > Out of curiosity, is there any way to access that log without being a part of Microsoft? The tests passed locally but I saw the bot report a failure and I couldn't figure out how to access the log myself.
> 
> Are you not able to access that link? I pasted the failures here anyway - https://gist.github.com/radical/1f55987361a1a828b3ed888b081e2dbe .

I can access the link you provided, but is there any way to access those logs from the failed check itself (e.g. assuming I don't have access to the azdo environment)?

**Comment by radical** [comment]
> < snip />
> I can access the link you provided, but is there any way to access those logs from the failed check itself (e.g. assuming I don't have access to the azdo environment)?

When the checks complete then you can see the details, and links to the failures in `Build Analysis`  under `Checks`.

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        clientBuilder.ConfigureOptions(options => configuration.Bind(options));
+#pragma warning restore IDE0200
+    }
+
+    protected override IAzureClientBuilder<EventHubBufferedProducerClient, EventHubBufferedProducerClientOptions> AddClient(AzureClientFactoryBuilder azureFactoryBuilder,
+        AzureMessagingEventHubsBufferedProducerSettings settings, string connectionName, string configurationSectionName)
+    {
+        return ((IAzureClientFactoryBuilderWithCredential) azureFactoryBuilder)
```
Comment (radical) [inline src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:30]: Should this confirm that the type is `IAzureClientFactoryBuilderWithCredential`?

**Review by radical — APPROVED** [review]
Other than the one comment, LGTM!

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs (hunk 5 above)
Comment (WhitWaldo) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubBufferedProducerClientComponent.cs:30]: The type provided is an `AzureClientFactoryBuilder` which implements `IAzureClientFactoryBuilderWithConfiguration<IConfiguration>` and `IAzureClientFactoryBuilderWithCredential`, so I wouldn't think so. Further, this pulled from the existing [EventHubConsumerClientComponent](https://github.com/WhitWaldo/dotnet-aspire/blob/2e968b6ae9771f294aca866f94179b176599ba00/src/Components/Aspire.Azure.Messaging.EventHubs/EventHubConsumerClientComponent.cs#L31)

I think if there's interest in doing some sort of full-project workover for best practices, that's probably better done in a separate PR as this one is seeking to minimally add a compliant implementation of the buffered producer client in the same approach already used for the already-provided other clients.

**Pull request merged**